### PR TITLE
feat(changelogs): prioritize markdown format for multiple file candidates

### LINKED ISSUE #25830: Renovate is using `CHANGELOG.json` when `CHANGELOG.md` also exists
### Discussed in https://github.com/renovatebot/renovate/discussions/25828

<div type='discussions-op-text'>

<sup>Originally posted by **frantic1048** November 16, 2023</sup>
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.8

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

## Symptom

We have some npm packages managed with beachball[^1], for each release, it generates both formats of changelog under package path:

- `CHANGELOG.json`: JSON file, changelog data for programmatical usage
- `CHANGELOG.md`: Markdown file, human readable changelog

For example: https://github.com/RightCapitalHQ/frontend-style-guide/tree/3fe7eea47b45ecd796340dc5ef1f98d034fd1c25/packages/eslint-config-javascript

When Renovate is generating Pull Request for such package, Renovate will report:

```
DEBUG: Multiple candidates for changelog file, using packages/eslint-config-javascript/CHANGELOG.json (repository=frantic1048/renovate-uses-json-changelog-while-md-exists, branch=action-renovate/rightcapital-eslint-config-javascript-14.x)
```

And generates empty release notes for the package. https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/pull/2

Example:

- Repo: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists
- Renovate config: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/renovate.json
- Renovate self-hosted config: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/renovate-config.js
- Renovate usage: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/.github/workflows/main.yml#L8
- Renovate Log: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/actions/runs/6888285336/job/18736975879#step:3:537

## Cause

In [`github.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L22) (called from [`getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/release-notes.ts#L237)), Renovate is [selecting the first file](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L63) filtered by [`changelogFilenameRegex`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L51), when multiple capable changelog files are found.

Unfortunately, `CHANGELOG.md` and `CHANGELOG.json` both pass the test of `changelogFilenameRegex`, resulting above symptom.

## Possible fix

Given the fact that Renovate is trying to [read changelog as a Markdown file](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/release-notes.ts#L216) to extract useful info.

I think Renovate may give markdown file a higher priority during selecting a changelog from multiple choices.

## Other info

[`gitlab.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/gitlab/index.ts#L17) and [`bitbucket.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/b3424c626cb8e44b922b91f669826a466ccee5c8/lib/workers/repository/update/pr/changelog/bitbucket/index.ts#L17) have similar logic and could suffer from same symptom 🤔 

[^1]: a tool to automating version bump and package publish for node projects, see: https://microsoft.github.io/beachball/

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Multiple candidates for changelog file, using packages/eslint-config-javascript/CHANGELOG.json (repository=frantic1048/renovate-uses-json-changelog-while-md-exists, branch=action-renovate/rightcapital-eslint-config-javascript-14.x)
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above</div>

### PR DESCRIPTION
**Pull request by frantic1048**
## Changes

1. Add a `sort()` before selecting changelog file in:
    1. `github.getReleaseNotesMd()`
    2. `gitlab.getReleaseNotesMd()`
    3. `bitbucket.getReleaseNotesMd()`

## Context

- resolves #25830

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by frantic1048** [comment]
[rebased](https://github.com/renovatebot/renovate/compare/bdf2058ed30ee87cee7d05160227078e7d96bbd9..69eed9b96173b69ad13cfb6ca26b25b5f217817e) to fix the conflict with current main branch.

File: lib/workers/repository/update/pr/changelog/common.ts [hunk 1]
```diff
@@ -5,3 +5,37 @@ export function slugifyUrl(url: string): string {
...
+ *  'CHANGELOG',
+ *  'CHANGELOG.json',
+ * ]
+ * ```
+ *
+ * @param a path to changelog file
+ * @param b path to changelog file
+ */
+export function compareChangelogFilePath(a: string, b: string): number {
+  const preferedChangelogRegexList = [/\.(md|markdown|mkd)$/i, /\.(txt|text)/i];
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/common.ts:31]:
```suggestion
  const preferedChangelogRegexList = [/\.(?:md|markdown|mkd)$/i, /\.(?:txt|text)$/i];
```

File: lib/workers/repository/update/pr/changelog/gitlab/index.ts [hunk 2]
```diff
@@ -16,7 +17,7 @@ const http = new GitlabHttp(id);
 export async function getReleaseNotesMd(
   repository: string,
   apiBaseUrl: string,
-  sourceDirectory?: string,
+  sourceDirectory?: string
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/gitlab/index.ts:20]: revert

File: lib/workers/repository/update/pr/changelog/gitlab/index.ts [hunk 3]
```diff
@@ -30,7 +31,7 @@ export async function getReleaseNotesMd(
       }`,
       {
         paginate: true,
-      },
+      }
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/gitlab/index.ts:34]: revert

File: lib/workers/repository/update/pr/changelog/gitlab/index.ts [hunk 4]
```diff
@@ -42,11 +43,13 @@ export async function getReleaseNotesMd(
...
   }
-  const { path: changelogFile, id } = files.shift()!;
+  const { path: changelogFile, id } = files
+    .sort((a, b) => compareChangelogFilePath(a.name, b.name))
+    .shift()!;
   /* istanbul ignore if */
   if (files.length !== 0) {
     logger.debug(
-      `Multiple candidates for changelog file, using ${changelogFile}`,
+      `Multiple candidates for changelog file, using ${changelogFile}`
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/gitlab/index.ts:52]: revert

File: lib/workers/repository/update/pr/changelog/gitlab/index.ts [hunk 5]
```diff
@@ -58,7 +61,7 @@ export async function getReleaseNotesMd(
 
 export async function getReleaseList(
   project: ChangeLogProject,
-  _release: ChangeLogRelease,
+  _release: ChangeLogRelease
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/gitlab/index.ts:64]: revert

**Review by viceice — COMMENTED** [review]
missing coverage

https://github.com/renovatebot/renovate/actions/runs/6913434308/job/18810264666#step:7:1

https://github.com/renovatebot/renovate/runs/18811225645

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**